### PR TITLE
implement stream based NAL parser and VideoStreamDecoder

### PR DESCRIPTION
**Pull request by mixaal**
This PR addresses: https://github.com/ralfbiedert/openh264-rs/issues/60

### CODE REVIEW

**Comment by ralfbiedert** [comment]
Thanks for the PR. 

I think the nal helper is generally OK, but I'd remove the decoder helper and its example since it's rather unrelated. 

Ill merge and do a bit clean up, but feel free to comment if you disagree.

**Comment by ralfbiedert** [comment]
Alright, I did some cleanup. 

Notably I removed the enum because in the current API I didn't find it helpful, and rather preferred an `Option` that just gives a result or tells me to keep going, similar to `nal_units()`. 

However, I see a potential issue with leftover NAL units. Right now you can't really differentiate them from incomplete packages. 

It could be an idea to bring that enum back, but with two changes : 

- All returned results should be slices to avoid the allocation / memcpy 
- Partial and / or the end NAL can be observable. 

Like so (naming TBD):

```rust
enum ParsingResult {
    None,
    Complete(&[u8]),
    PossiblyPartial(&[u8]) 
}
```

Totally unrelated, I saw you provide a method (now removed):

```rust
    fn make<T>(capacity: usize) -> Vec<T> {
        let mut v = Vec::with_capacity(capacity);
        unsafe {
            v.set_len(capacity);
        }
        v
    }
```
You should never do that as this is undefined behavior; the canonical way would just be `vec![0; N]` to create a vector that has `N` elements.

**Comment by mixaal** [comment]
I think you left video_stream as an example, however it has imports from removed file stream.rs. Maybe you want to remove this example?

**Comment by mixaal** [comment]
There is one thing - replacing the enum with the Option will feed the buffer without the possibility to scan through the buffer, i.e. what CallNext vs ReadMore in the original code was supposed to do. If we have 1M stream with 8k packets we read much more than we are able to parse. I think it might be beneficial to return some sort of  distinction between CallNext vs ReadMore - whatever the naming is. Thanks.

**Comment by ralfbiedert** [comment]
> I think you left video_stream as an example, however it has imports from removed file stream.rs. Maybe you want to remove this example? 

Ah yes, thanks!

>  without the possibility to scan through the buffer, i.e. what CallNext vs ReadMore in the original code was supposed to do

I'm open to adding some of the removed flexibility, I mainly felt the code was not as idiomatic and a bit hard to follow, so I wanted to start with a reduced API surface I could wrap my head around.  To improve the current state could we vaguely in this order:

- Get rid of the `Vec<u8>`, because most other parts of the API are rather conservative about allocation (e.g., `nal_units()`)
- Possibly reuse `nal_units()` to simplify the scanning logic (if that doesn't make sense perf. wise feel free to skip)
- Make a more idiomatic proposal how that scanning behavior could look like instead
- Prevent the `Vec` from growing indefinitely 

Since I don't have a use case for this I'm not fixated on having the enum I mentioned above (although it would be nice to have from an API perspective), but from a maintainability and performance perspective I'd prefer to streamline this somewhat according to those points.

I happily help with a PR for that.
